apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  SemaphoreIdentityPCD,
+  SemaphoreIdentityPCDPackage
+} from "@pcd/semaphore-identity-pcd";
+import { Identity } from "@semaphore-protocol/identity";
+import { appConfig } from "./appConfig";
+import { loadSelf } from "./localstorage";
+
+/**
+ * Validates the application state using {@link validateAppState}. In the case
+ * that the state is invalid, returns `true`, and concurrently uploads the validation
+ * errors to the server for further inspection.
+ *
+ * In the case there are no validation errors, returns `false`.
+ */
+export function validateAndLogStateErrors(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): boolean {
+  const validationErrors = validateAppState(
+    self,
+    identity,
+    pcds,
+    forceCheckPCDs
+  );
+
+  if (validationErrors.errors.length > 0) {
+    logValidationErrors(validationErrors);
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Determines whether the app's global state contains valid data. If it does not,
+ * returns the set of things that are incorrect about it in a {@link ValidationErrors}
+ * object. If there were no validation errors the result will contain an empty array
+ * of errors.
+ *
+ * The provided {@link PCDCollection} is not checked unless either this function
+ * determines the user is logged in or the {@link forceCheckPCDs} argument is `true`.
+ */
+function validateAppState(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): ValidationErrors {
+  const validationErrors: string[] = [];
+
+  const loggedOut = !self;
+
+  const identityPCDFromCollection = pcds?.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )?.[0] as SemaphoreIdentityPCD | undefined;
+
+  if (forceCheckPCDs || !loggedOut) {
+    if (!identityPCDFromCollection) {
+      validationErrors.push(
+        "'pcds' field in app state does not contain an identity PCD"
+      );
+    }
+  }
+
+  if (loggedOut) {
+    return {
+      errors: validationErrors,
+      userUUID: undefined
+    };
+  }
+
+  if (!self) {
+    validationErrors.push("missing 'self'");
+  }
+
+  if (!identity) {
+    validationErrors.push("missing 'identity'");
+  }
+
+  if (!pcds) {
+    validationErrors.push("missing 'pcds'");
+  }
+
+  if (pcds?.size() === 0) {
+    validationErrors.push("'pcds' contains no pcds");
+  }
+
+  const identityFromPCDCollection = identityPCDFromCollection?.claim?.identity;
+  const commitmentOfIdentityPCDInCollection =
+    identityFromPCDCollection?.commitment?.toString();
+  const commitmentFromSelfField = self?.commitment;
+  const commitmentFromIdentityField = identity?.commitment?.toString();
+
+  if (commitmentOfIdentityPCDInCollection !== commitmentFromSelfField) {
+    validationErrors.push(
+      `commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})` +
```

Comment:
To be explicit, the full set of user data potentially included in these logs is UUID and semaphore commitment.  Both of those are already known and correlated on the server (both are in the User object), so I see no privacy issues.